python_modules/dagster/dagster/_core/definitions/data_version.py

```diff
@@ -503,21 +621,29 @@ def _is_partitioned_or_downstream(self, *, key: "AssetKey") -> bool:
                 for dep_key in self.asset_graph.get_parents(key)
             )
 
+    @cached_method
+    def _get_materialization_count_by_partition(self, *, key: "AssetKey", cursor: int):
+        return self._instance.get_materialization_count_by_partition([key], cursor)
+
     # Volatility means that an asset is assumed to be constantly changing. We assume that observable
     # source assets are non-volatile, since the primary purpose of the observation function is to
     # determine if a source asset has changed. We assume that regular assets are volatile if they
     # are at the root of the graph (have no dependencies) or are downstream of a volatile asset.
     @cached_method
-    def _is_volatile(self, *, key: "AssetKey") -> bool:
-        if self.asset_graph.is_source(key):
-            return self.asset_graph.is_observable(key)
+    def _is_volatile(self, *, key: "AssetKeyPartitionKey") -> bool:
+        from dagster._core.definitions.events import AssetKeyPartitionKey
+
+        if self.asset_graph.is_source(key.asset_key):
+            return self.asset_graph.is_observable(key.asset_key)
         else:
-            deps = self.asset_graph.get_parents(key)
+            deps = [
```

Comment:
I don't think modifying this method is necessary -- volatility is still linked to the asset key, and so we shouldn't need to traverse the full-resolution asset graph.

Reply:
There was no real change to the logic here, I had changed it just so that all the cached methods use `AssetKeyPartitionKey` to keep things simple, but your comment has made me realize that there is a performance cost for multiple partitions. Reverted.